File: cs/UCLA/libUCLA/ULoader_JSON.cs
using libUCLA;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.IO;

namespace libUCLA {

    /// <summary>
    /// This exception is being thrown when something go wrong during config processing.
    /// For best debugging, you should check inner exception.
    /// </summary>
    [Serializable]    
    public class ULoaderException : UException {
        public ULoaderException() : base() { }
        public ULoaderException(string message) : base(message) { }
        public ULoaderException(string message, Exception innerException) : base(message, innerException) { }
    }

    public class ULoader_JSON {
        private JObject outputConfig;
        private JObject inputConfig;

        /// <summary>
        /// Preloads configuration file.
        /// </summary>
        /// <param name="configFile">File name</param>
        public ULoader_JSON(string configFile) {
            try {
                using (StreamReader reader = File.OpenText(configFile)) {
                    JObject config = (JObject) JToken.ReadFrom(new JsonTextReader(reader));

                    //Inputs
                    if (config.GetValue("inputs") != null) {
                        var inputs = config.GetValue("inputs").ToString(Newtonsoft.Json.Formatting.None);
                        this.inputConfig = JObject.Parse(inputs);
                    }

                    //Outputs
                    if (config.GetValue("outputs") != null) {
                        var outputs = config.GetValue("outputs").ToString(Newtonsoft.Json.Formatting.None);
                        this.outputConfig = JObject.Parse(outputs);
                    }
                }
            } catch (Exception ex) {
                throw new ULoaderException("Cannot load configuration from file: '" + configFile + "'", ex);
            }
        }

        /// <summary>
        /// Creates configuration object for receiver.
        /// </summary>
        /// <param name="inputName">Input name.</param>
        /// <returns>Configuration object.</returns>
        public UConfig GetReceiverConfig(string inputName) {
            if (this.inputConfig.GetValue(inputName) == null)
                throw new ULoaderException("There is no such input defined in config file. Input name: '" + inputName + "'");

            var configInterface = this.inputConfig.GetValue(inputName)["interface"];
            if (configInterface == null)
                throw new ULoaderException("Cannot read 'interface' vale from specified configuration.");

            var configPort = this.inputConfig.GetValue(inputName)["port"];
            if (configPort == null)
                throw new ULoaderException("Cannot read 'port' vale from specified configuration.");

            UConfig config = new UConfig((String) configInterface, (String) configPort);

            return config;
        }

        /// <summary>
        /// Creates receiver based on configuration.
        /// </summary>
        /// <param name="inputName">Name of input.</param>
        /// <returns>Receiver object. (no autostart)</returns>
        public UReceiver GetReceiver(string inputName) {
            UReceiver server = new UReceiver(GetReceiverConfig(inputName));
            return server;
        }

        /// <summary>
        /// Creates configuration object for sender.
        /// </summary>
        /// <param name="outputName">Output name.</param>
        /// <returns>Configuration object.</returns>
        public UConfig GetSenderConfig(string outputName) {
            if (this.outputConfig.GetValue(outputName) == null)
                throw new ULoaderException("There is no such output defined in config file. Output name: '" + outputName + "'");

            var configHost = this.outputConfig.GetValue(outputName)["host"];
            if (configHost == null)
                throw new ULoaderException("Cannot read 'host' vale from specified configuration.");

            var configPort = this.outputConfig.GetValue(outputName)["port"];
            if (configPort == null)
                throw new ULoaderException("Cannot read 'port' vale from specified configuration.");

            UConfig config = new UConfig((String) configHost, (String) configPort);
            return config;
        }

        /// <summary>
        /// Creates sender based on configuration.
        /// </summary>
        /// <param name="outputName">Name of output.</param>
        /// <returns>Sender object. (no autostart)</returns>
        public USender GetSender(string outputName) {
            USender client = new USender(GetSenderConfig(outputName));
            return client;
        }
    }
}

File: cpp/UCLA/libUCLA/ULoader_JSON.cpp
#include "ULoader_JSON.h"
#include <jansson.h>

using namespace UCLA;

ULoader_JSON::ULoader_JSON(char *configFile){
	json_t *config;
	json_error_t error;
	json_t *inputs, *outputs;
	this->receiverConfig = NULL;
	this->senderConfig = NULL;

	try{
		config = json_load_file(configFile, 0, &error);
		if(config) {
			//Inputs
			inputs = json_object_get(config, "inputs");
			if(json_is_object(inputs))
			{
				this->receiverConfig = inputs;
			}

			//Outputs
			outputs = json_object_get(config, "outputs");
			if(json_is_object(outputs))
			{
				this->senderConfig = outputs;
			}

		}else
			throw std::exception();

	} catch(...){
		throw UException("Cannot load configuration from file");
	}      
}

std::unique_ptr<UConfig> ULoader_JSON::GetReceiverConfig(char *inputName) {
	json_t *server, *configInterface, *configPort;

	server = json_object_get(static_cast<json_t*>(this->receiverConfig), inputName);
	if(!json_is_object(server))	
		throw UException("There is no such server defined in config file.");

	configInterface = json_object_get(server, "interface");
	if(!json_is_string(configInterface))	
		throw UException("Cannot read 'interface' value from specified configuration.");

	configPort = json_object_get(server, "port");
	if(!json_is_string(configPort))	
		throw UException("Cannot read 'port' value from specified configuration.");

	std::unique_ptr<UConfig> config(new UConfig(json_string_value(configInterface), json_string_value(configPort)));

	return std::move(config);
}

std::unique_ptr<UReceiver> ULoader_JSON::GetReceiver(char *inputName, std::function<void(char*, int)> handler) {
	std::unique_ptr<UConfig> config = GetReceiverConfig(inputName);
	std::unique_ptr<UReceiver> server(new UReceiver(*config, handler));	

	return std::move(server);
}

std::unique_ptr<UConfig> ULoader_JSON::GetSenderConfig(char *outputName) {
	json_t *server, *configHost, *configPort;

	server = json_object_get(static_cast<json_t*>(this->senderConfig), outputName);
	if(!json_is_object(server))	
		throw UException("There is no such client defined in config file.");

	configHost = json_object_get(server, "host");
	if(!json_is_string(configHost))	
		throw UException("Cannot read 'host' value from specified configuration.");

	configPort = json_object_get(server, "port");
	if(!json_is_string(configPort))	
		throw UException("Cannot read 'port' value from specified configuration.");

	std::unique_ptr<UConfig> config(new UConfig(json_string_value(configHost), json_string_value(configPort)));

	return std::move(config);
}

std::unique_ptr<USender> ULoader_JSON::GetSender(char *outputName) {
	std::unique_ptr<UConfig> config = GetSenderConfig(outputName);
	std::unique_ptr<USender> client(new USender(*config));	

	return std::move(client);
}
